### PR TITLE
ci: GitHub Actions

### PR DESCRIPTION
**Pull request by ahartikainen**
Run tests in GitHub Actions, testing against Linux (Ubuntu 20.04) and macOS (10.14).

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
+name: pystan
+
+on:
+  push:
+    branches:
+      - "**"
+    tags:
+      - "*.*.*"
+  pull_request:
```
Comment (ahartikainen) [inline .github/workflows/main.yml:9]: This was harder to set up.

Basically `[push, pull_request]` should work but then the `if` should extract correct format in publish job.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+jobs:
+  tests:
+    name: pystan tests
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ubuntu-20.04, macos-10.14]
```
Comment (ahartikainen) [inline .github/workflows/main.yml:16]: What image should we use? Do we follow httpstan?

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      - name: Check out repository
+        uses: actions/checkout@v2
+
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: Install poetry and tox
+        run: pip install "poetry==1.0.10" tox
```
Comment (ahartikainen) [inline .github/workflows/main.yml:28]: httpstan uses `1.0.9`, The publish part used `1.0.10` and travis had `pip install "poetry<2,>=1.0"`

File: .github/workflows/main.yml [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        if: matrix.python-version == '3.8' && matrix.os == 'ubuntu-20.04'
+        run: tox
+
+      - name: Build and Install wheel
+        run: |
+          poetry build -v
+          python -m pip install dist/*.whl
+
+      - name: Run tests
+        run: python -m pytest -s -v --cov=stan tests
```
Comment (ahartikainen) [inline .github/workflows/main.yml:46]: I follow httpstan structure here.

Travis has 

    poetry install -v && poetry run pytest -s -v tests

Which one we do want to follow?

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  publish:
+    name: publish to pypi
+    runs-on: ubuntu-20.04
+    needs: tests
+    if: startsWith(github.ref, 'refs/tags/')
```
Comment (ahartikainen) [inline .github/workflows/main.yml:52]: This first checks that tests are passed and then checks if the trigger came from a tag.

File: .github/workflows/main.yml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        run: |
+          poetry build -v
+          python -m pip install dist/*.whl
+
+      - name: Run tests
+        run: python -m pytest -s -v --cov=stan tests
+
+  publish:
+    name: publish to pypi
+    runs-on: ubuntu-20.04
```
Comment (ahartikainen) [inline .github/workflows/main.yml:50]: Here also, what image should we use

File: .github/workflows/main.yml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    needs: tests
+    if: startsWith(github.ref, 'refs/tags/')
+    steps:
+      - name: Check out repository
+        uses: actions/checkout@v2
+
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.x"
```
Comment (ahartikainen) [inline .github/workflows/main.yml:60]: This uses the latest stable?

**Comment by ahartikainen** [comment]
Here is one version of GH Actions workflow. It is not 100% the same as the travis file. I added some comments.

**Comment by ahartikainen** [comment]
Also, macOS tests fail randomly

```
__________________________ test_fit_cache_uncacheable __________________________

    def test_fit_cache_uncacheable():
        """Test that a fit with a random seed set is cached."""
        cache_size_before = shutil.disk_usage(cache_path()).used
        # this fit is NOT cacheable, should not be saved
        normal_posterior = stan.build(program_code)
        normal_posterior.sample()
        cache_size_after = shutil.disk_usage(cache_path()).used
        # allow for a 4096 byte difference (an empty directory takes 4K)
>       assert abs(cache_size_before - cache_size_after) <= 4096
E       assert 348160 <= 4096
E        +  where 348160 = abs((284819333120 - 284818984960))

tests/test_fit_cache.py:45: AssertionError
```

**Comment by riddell-stan** [comment]
On the wheel publishing, I think the practice is to make that a separate workflow. (Travis CI does not have a concept of separate workflows.)

Perhaps one PR could introduce testing and a second PR introduce the wheel publishing workflow.

Sorry to hear that macos builds are failing randomly. We'll have to solve that problem before we can switch.

File: .github/workflows/main.yml (hunk 2 above)
Comment (riddell-stan) [reply .github/workflows/main.yml:16]: Yes, that sounds right to me.

File: .github/workflows/main.yml (hunk 4 above)
Comment (riddell-stan) [reply .github/workflows/main.yml:46]: We can use a simpler, more standard step here. PyStan does not have a `Makefile` or a `build.py` that we need to worry about. So I'd keep: `poetry install -v && poetry run pytest -s -v tests`

File: .github/workflows/main.yml (hunk 7 above)
Comment (riddell-stan) [reply .github/workflows/main.yml:60]: Let's narrowly copy the Travis CI config. So 3.7 and 3.8 for now. Adding 3.9 can be a separate PR once everything works.

**Comment by riddell-stan** [comment]
This can be two commits, one which switches to github actions and another one which fixes the flaky test.

The commit that's listed right now is incorrect because it doesn't mention switching to github actions.

File: .github/workflows/main.yml (hunk 3 above)
Comment (riddell-stan) [reply .github/workflows/main.yml:23]: Let's stick with the original "poetry<2,>=1.0". The version that httpstan uses shouldn't matter here. We don't depend on poetry for anything other than building.

**Comment by ahartikainen** [comment]
@riddell-stan I added some prints to the cache test (the commit message is a bit wrong, but we can remove that commit after we figure out how to fix this)

In Ubuntu

```
tests/test_fit_cache.py::test_fit_cache 60187803648
Files in cache (before): ['xde6st3p', '4jtrdzjj', 'npn4yfou', 'ubsfyorz', 'e6jggr4p', 'arqv5ke7', 'm6orjnfj', 'fst3i4w4']
60187987968
Files in cache (while): ['xde6st3p', '4jtrdzjj', 'npn4yfou', 'ubsfyorz', 'e6jggr4p', 'arqv5ke7', 'm6orjnfj', 'fst3i4w4']
Files in cache (after): ['xde6st3p', '4jtrdzjj', 'npn4yfou', 'ubsfyorz', 'e6jggr4p', 'arqv5ke7', 'm6orjnfj', 'fst3i4w4']
PASSED
```

In macOS
```
tests/test_fit_cache.py::test_fit_cache 286168182784
Files in cache (before): ['6pwjchmp', 'dfazcbbz', 'ycrtzmop', 'prfylupj', '65zrwsed', '2twplmj6', '4b67nwn7', 'j3peqnxo']
286168367104
Files in cache (while): ['6pwjchmp', 'dfazcbbz', 'ycrtzmop', 'prfylupj', '65zrwsed', '2twplmj6', '4b67nwn7', 'j3peqnxo']
Files in cache (after): ['6pwjchmp', 'dfazcbbz', 'ycrtzmop', 'prfylupj', '65zrwsed', '2twplmj6', '4b67nwn7', 'j3peqnxo']
PASSED
```

So, I think cache test is a bit flawed? The cache should be empty, right?

**Comment by riddell-stan** [comment]
The tests seem to be passing. We would expect this, right? The GitHub Actions is a straight copy of the Travis CI config, which also yielded passing tests.

If the tests pass, let's merge this (without the debug printing). If there appears to be a flaky test, we can create a new issue.

**Comment by ahartikainen** [comment]
Yes, let's do that.


My point was that given the following structure

```python
    cache_size_before = shutil.disk_usage(cache_path()).used
    print(cache_size_before)
    print("Files in cache (before):", [item.name for item in cache_path().glob("*")])
    # this fit is cacheable
    random_seed = random.randrange(1, 2 ** 16)
    normal_posterior = stan.build(program_code, random_seed=random_seed)
    normal_posterior.sample()
    cache_size_after = shutil.disk_usage(cache_path()).used
    print(cache_size_after)
    print("Files in cache (while):", [item.name for item in cache_path().glob("*")])
    assert cache_size_after > cache_size_before

    # fit is now in cache
    cache_size_before = shutil.disk_usage(cache_path()).used
    normal_posterior.sample()
    cache_size_after = shutil.disk_usage(cache_path()).used
    print("Files in cache (after):", [item.name for item in cache_path().glob("*")])
```

There should be a new fit in cache, but either it doesn't create a new fit, or the "new" fit is already cached.

One reason for this could be that we set a random seed somewhere (??) so the random number is always the same --> the fit already exists. This might be something we / GH_Actions do.

**Comment by riddell-stan** [comment]
re:

```
print("Files in cache (before):", [item.name for item in cache_path().glob("*")])
```
Isn't this perhaps only showing cached models, not cached fits? Fits are stored in their "parent" models' directories.

**Comment by ahartikainen** [comment]
Of course. Yes.

**Comment by ahartikainen** [comment]
Ok, now it fails again

```
>       assert abs(cache_size_before - cache_size_after) <= 4096
E       assert 16384 <= 4096
E        +  where 16384 = abs((286331031552 - 286331015168))
```

**Comment by riddell-stan** [comment]
Let's change 4096 to 16384. That seems like the fastest fit. Maybe adjust the comment a bit, "allow for empty directories (4096 bytes on Linux) and other chance file creations."

Small fits are at least 100K, so this doesn't invalidate the test.

**Comment by riddell-stan** [comment]
@ahartikainen looks like you're busy. Since #173 depends on this getting fixed, I'll try to work on this tomorrow (Wed) in a separate PR.

**Comment by ahartikainen** [comment]
Hi,

I can push the change for the test.

Problem is, that I have done a bit more digging and this cache error also happens on Ubuntu.

It is probably some error log file or something similar. I'm trying to trigger it again to see what is going on.

**Review by riddell-stan — APPROVED** [review]

**Comment by riddell-stan** [comment]
One of the test jobs got stuck (not on a cache issue).

I'll merge this in the hope that we can figure out the issues quickly. We know that Github Actions *can* work because it's working on httpstan right now.

**Pull request merged**